server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -572,4 +576,45 @@ private void update(DbSession session, RuleDefinitionDto rule) {
     dbClient.ruleDao().update(session, rule);
   }
 
+
+  private static void verifyRuleKeyConsistency(RulesDefinition.Context context) {
+    List<RulesDefinition.Rule> allRules = context.repositories().stream()
+      .flatMap(r -> r.rules().stream())
+      .collect(MoreCollectors.toList());
+
+    List<RuleKey> definedRuleKeys = allRules.stream()
+      .map(r -> RuleKey.of(r.repository().key(), r.key()))
+      .collect(Collectors.toList());
+
+    List<RuleKey> definedDeprecatedRuleKeys = allRules.stream()
+      .flatMap(r -> r.deprecatedRuleKeys().stream())
+      .collect(Collectors.toList());
+
+    Set<RuleKey> duplicates = findDuplicates(definedDeprecatedRuleKeys);
+    if (!duplicates.isEmpty()) {
```

Comment:
you can use `Preconditions.checkState`:
* get rid of if condition
* get rid of String.format
* get rid of `throw new IllegalStateException`